annnufan/assignment1/solution_1.cpp

```diff
@@ -0,0 +1,104 @@
+#include <iostream>
+#include <string>
+#include <cstring>
+#include <vector>
+#include <algorithm>
+#include <cctype>
+#include <functional>
+
+void word_anagram(bool case_sens) {
+	std::string word_1, word_2;
+	std::cin >> word_1 >> word_2;
+	if (!case_sens) {
+		std::transform(word_1.begin(), word_1.end(), word_1.begin(), std::ptr_fun <int, int> (std::tolower));
+		std::transform(word_2.begin(), word_2.end(), word_2.begin(), std::ptr_fun <int, int> (std::tolower));
+	}
+	if (word_1.length() != word_2.length()) {
+		std::cout << "Second word isn't anagram of first\n";
+		return;
+	}
+
+	std::vector<int> count(256, 0);
+	for (auto ch : word_1) {
+		count[ch]++;
+	}
+	for (auto ch : word_2) {
+		if (count[ch] > 0) {
+			count[ch]--;
+		} else {
+			std::cout << "Second word isn't anagram of first\n";
+			return;
+		}
+	}
+
+	std::cout << "Yes, second word is anagram of first\n";
+	return;
+}
+
+std::vector<std::string> line_to_words(std::string line) {
+	std::string word = "";
+	std::vector<std::string> words;
+	for (int i = 0; i < line.length(); i++) {
+		if (line[i] == ' ' ) {
+			if (word != "") {
+				std::sort(word.begin(), word.end());
+				words.push_back(word);
+				word = "";
+			}
+		} else {
+			word += line[i];
+		}
+	}
+	if (word != "") {
+		std::sort(word.begin(), word.end());
+		words.push_back(word);
+	}
+	return words;
+}
+
+void sentence_anagram(bool case_sens) {
+	std::string line_1, line_2;
+	std::getline(std::cin, line_1);
+	std::getline(std::cin, line_1);
+	std::getline(std::cin, line_2);
+	if (!case_sens) {
+		std::transform(line_1.begin(), line_1.end(), line_1.begin(), std::ptr_fun <int, int>(std::tolower));

```

Comment:
No need to `ptr_fun` here, std::transform works with a template predicate, meaning you can pass anything with a matching `operator()` including native functions.
```C++
std::transform(line_1.begin(), line_1.end(), line_1.begin(), std::tolower);
```

Also, `ptr_fun` is deprecated according to [cppreference](http://en.cppreference.com/w/cpp/utility/functional/ptr_fun).

Reply:
If I just write ```std::transform(word_1.begin(), word_1.end(), word_1.begin(), std::tolower);``` it isn't compile.